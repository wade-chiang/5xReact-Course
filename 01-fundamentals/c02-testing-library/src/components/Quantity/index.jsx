import { memo } from 'react';

/* type QuantityProps = {
  value?: number,
  maxValue?: number,
  onChange: (quantity: number) => void,
}; */

const Quantity = (props) => {
  const { value = 0, maxValue = Infinity, onChange } = props;

  const atDecrease = () => onChange(value - 1);
  const atIncrease = () => onChange(value + 1);

  return (
    <section data-name="Quantity" data-max-value={maxValue}>
      <button onClick={atDecrease} disabled={value <= 0}>
        -
      </button>
      <div data-testid="quantity">{value}</div>
      <button onClick={atIncrease} disabled={value >= maxValue}>
        +
      </button>
    </section>
  );
};

export default memo(Quantity);
