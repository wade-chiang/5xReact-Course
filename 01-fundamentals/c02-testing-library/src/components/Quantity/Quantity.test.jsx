import { render, fireEvent } from '@testing-library/react';
import Quantity from '.';

describe('<Quantity />', () => {
  test('should be render correct', () => {
    const { getByTestId, container } = render(<Quantity />);
    expect(getByTestId('quantity')).toHaveTextContent('0');

    expect(container).toMatchSnapshot();
  });

  test('should display correct quantity when props change', () => {
    const { getByTestId, rerender, container } = render(<Quantity />);

    rerender(<Quantity value={10} />);
    expect(getByTestId('quantity')).toHaveTextContent('10');
    expect(container).toMatchSnapshot();
  });

  test('should toggle callback when + button click', () => {
    const onChange = vi.fn();
    const { getByText } = render(<Quantity onChange={onChange} />);

    fireEvent.click(getByText('+'));
    expect(onChange).toBeCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toBe(1);
  });

  test('should disable button when value > maxValue', () => {
    const { getByText, rerender } = render(<Quantity />);

    expect(getByText('-')).toBeDisabled();
    expect(getByText('+')).not.toBeDisabled();

    rerender(<Quantity value={5} maxValue={5} />);
    expect(getByText('-')).not.toBeDisabled();
    expect(getByText('+')).toBeDisabled();
  });
});
