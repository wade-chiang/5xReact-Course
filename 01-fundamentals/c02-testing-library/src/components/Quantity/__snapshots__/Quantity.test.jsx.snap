// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Quantity /> > should be render correct 1`] = `
<div>
  <section
    data-max-value="Infinity"
    data-name="Quantity"
  >
    <button
      disabled=""
    >
      -
    </button>
    <div
      data-testid="quantity"
    >
      0
    </div>
    <button>
      +
    </button>
  </section>
</div>
`;

exports[`<Quantity /> > should display correct quantity when props change 1`] = `
<div>
  <section
    data-max-value="Infinity"
    data-name="Quantity"
  >
    <button>
      -
    </button>
    <div
      data-testid="quantity"
    >
      10
    </div>
    <button>
      +
    </button>
  </section>
</div>
`;

exports[`<Quantity /> should be render correct 1`] = `
<div>
  <section
    data-max-value="Infinity"
    data-name="Quantity"
  >
    <button
      disabled=""
    >
      -
    </button>
    <div
      data-testid="quantity"
    >
      0
    </div>
    <button>
      +
    </button>
  </section>
</div>
`;

exports[`<Quantity /> should display correct quantity when props change 1`] = `
<div>
  <section
    data-max-value="Infinity"
    data-name="Quantity"
  >
    <button>
      -
    </button>
    <div
      data-testid="quantity"
    >
      10
    </div>
    <button>
      +
    </button>
  </section>
</div>
`;
